[logUpload] Limit upload size

deprov447 opened the pull request:
Implementing a configuration key `sizeLimit` from `branding.desc` to limit maximum size of logs to be uploaded.

deprov447 committed:
[logUpload] Configurable upload size limit

A key 'sizeLimit' added to uploadServer field in branding.desc to limit the size of logFile to upload.

deprov447 committed:
[logUpload] fixes

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +40,15 @@ logFileContents()
     }
     QFileInfo fi( pasteSourceFile );
-    if ( fi.size() > 16_KiB )
+    sizeLimit = ( sizeLimit < 0 ) ? 1024*1024 : sizeLimit * 1024;            //For KiB to bytes
```
The documentation says "no limit", while this implementation turns it into 1024KiB.

deprov447 replied:
fixed. used `pasteSourceFile.readAll()`

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -30,7 +30,7 @@ using namespace CalamaresUtils::Units;
  */
 STATICTEST QByteArray
-logFileContents()
+logFileContents( qint64 sizeLimit )
```
Since this is intended to be in KiB, it's probably useful to name it `sizeLimitKiB` or so.

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +40,15 @@ logFileContents()
         return QByteArray();
     }
```
If `sizeLimit` is 0, you can also bail out early.

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +40,15 @@ logFileContents()
     }
     QFileInfo fi( pasteSourceFile );
-    if ( fi.size() > 16_KiB )
+    sizeLimit = ( sizeLimit < 0 ) ? 1024*1024 : sizeLimit * 1024;            //For KiB to bytes
```
Hm .. I was also going to suggest using (from `utils/Units.h`) function `KiBtoBytes()`, but that takes unsigned. You would need to cast `sizeLimit` to unsigned for that (which is ok: in that branch of the ternary, you know it is non-negative)

deprov447 replied:
You asked to convert KiB to bytes in `Calamares::Branding::uploadServerFromMap()` for _not having to_ care about units everytime UploadServerInfo object is made. But _utils_ in _libcalamaresui_ do not provide unit converter. So either I have to explicitly convert them (*1024) or call utils from _libcalamares_. I guess 1st one would be better ?

adriaandegroot replied:
`src/libcalamares/utils/Units.h` (due to the way include paths are, just `#include "utils/Units.h"` ), then `CalamaresUtils::Units::KiBtoBytes()`

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +40,15 @@ logFileContents()
+    if ( fi.size() > sizeLimit and sizeLimit > 0 )
     {
-        pasteSourceFile.seek( fi.size() - 16_KiB );
+        // Fixme : this following line is not getting pasted
```
Should be `FIXME` (until it's fixed :) )

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +40,15 @@ logFileContents()
     {
-        pasteSourceFile.seek( fi.size() - 16_KiB );
+        // Fixme : this following line is not getting pasted
+        cDebug() << "Only last " << sizeLimit << " bytes of log file (" << fi.size() << ") uploaded" ;
```
For debugging this (and the FIXME) you should write a static, C-style function that does a `stat()` on the log file (get its name via `Logger::logFile()`) and `fprintf(stderr, ..)` the results.

What I suspect is this: `QFileInfo` does the `stat()` call in its constructor; let's say it finds the file to be exactly 70000 bytes long. Now you log one extra line, so you have about 70060 bytes in the file. Below, we `seek()` to the size from the fileinfo, minus the size limit: but that is `70000 - sizeLimit`, not the **current** size of the file after that debug line! So you never see that line in the uploaded log, even though it might be there in the log at the moment of upload.

Another way of testing this is to seek to `fi.size()-sizeLimit+10`, or to read `sizeLimit+10`, just so see if the bytes are there. Come to think of it, reading `sizeLimit+1_KiB` might even be a good thing **anyway** in case new logging was done from another thread.

adriaandegroot replied:
You don't need extra spaces between items; `std::cout` does not introduce spaces between items, but QDebug does (by default, although there is a NoQuote item you can send to turn that off).

adriaandegroot commented:
I'm going to backpedal a little on some of my comments: I think the API should send around **bytes** for sizeLimit,  not KiB.

The reason for that can be seen in the unit tests:
```
QByteArray contentsOfLogfileAfterSetup = logFileContents( 16 );
```
16? 16 what? If we were to write 16_KiB it would be immediately clear (well, clearer that this is byte count).

The change that is needed for that is to move the convert-value-to-KiB into the part that reads the **branding** info and returns the `UploadServerInfo`. That is, in `uploadServerFromMap()`, take the size limit from the config file, and convert right there into bytes: -1 for negative values, 0 for 0, * 1024 (use `KiBtoBytes()`) for others. Then any consumers of an `UploadServerInfo` object don't need to worry about KiB-vs-anything else: it's just a byte count.

[[ Somewhat related: the docs for `UploadServerInfo` are still for the pair-valued one, not the 3-tuple. ]]

adriaandegroot commented:
Overall: looks pretty good, could be better in some spots, I support your wish for FIXME -- I think that you can hunt that down with some debugging experiments -- and a little bit of polish will make it just lovely.

deprov447 committed:
[logUpload] suggestionsAndFixes

deprov447 committed:
[logUpload] suggestionsAndFixes-part2

adriaandegroot commented on src/branding/default/branding.desc:
```diff
@@ -220,13 +220,17 @@ slideshowAPI: 2
+#                the scheme is not used.
+#  - sizeLimit : Defines maximum size limit (in KiB) of log file to be pasted. 
+#                Takes integer as input. If <=0, no limit will be forced,
+#                else only last 'n' KiB of log file will be pasted. 
```
The comment now says that 0 is "no limit", but the implementation uses 0 as "no data", so there's a mismatch.

adriaandegroot replied:
Probably also worthwhile mentioning that slightly more than the limit may be uploaded, due to last-minute logging.

adriaandegroot commented on src/libcalamaresui/Branding.h:
```diff
@@ -226,7 +226,7 @@ class UIDLLEXPORT Branding : public QObject
      * This is both the type (which may be none, in which case the URL
      * is irrelevant and usually empty) and the URL for the upload.
      */
```
Still needs documentation for that third element in the tuple (which you have kept as "number of kibibytes")

adriaandegroot commented on src/libcalamaresui/ViewManager.cpp:
```diff
@@ -144,7 +144,7 @@ void
 {
     bool shouldOfferWebPaste
-        = Calamares::Branding::instance()->uploadServer().first != Calamares::Branding::UploadServerType::None;
+        = std::get<0>(Calamares::Branding::instance()->uploadServer()) != Calamares::Branding::UploadServerType::None;
```
Depending on the meaning of `sizeLimit: 0` (according to the docs, or according to the implementation), you could also check that here.

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +42,16 @@ logFileContents()
     QFileInfo fi( pasteSourceFile );
-    if ( fi.size() > 16_KiB )
+    if( sizeLimitKiB < 0 )
+        sizeLimitKiB = 1024;
```
This deserves a test as well (in `TestPaste.cpp`)

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +42,16 @@ logFileContents()
-    if ( fi.size() > 16_KiB )
+    if( sizeLimitKiB < 0 )
+        sizeLimitKiB = 1024;
+    qint64 sizeLimitBytes = CalamaresUtils::KiBtoBytes( ( unsigned long long ) sizeLimitKiB );
```
Please use `static_cast<>` rather than C-style cast. The `sizeLimitBytes` can be const, as well.

adriaandegroot commented on src/libcalamaresui/utils/Paste.cpp:
```diff
@@ -40,11 +42,16 @@ logFileContents()
     {
-        pasteSourceFile.seek( fi.size() - 16_KiB );
+        cDebug() << "Only last" << sizeLimitBytes << "bytes of log file (sized" << fi.size() << "bytes) uploaded" ;
+        pasteSourceFile.seek( fi.size() - sizeLimitBytes + 1_KiB );
```
Changing the seek **and** changing the read (below) seems to defeat the purpose: suppose the log file was 64KiB and the sizeLimit is 16KiB, then:
- you seek to 49KiB
- if there's been no additional writes to the log file, you read 17KiB, and get 15KiB
- if there's been writes, less than 1KiB you read 17KiB and still get < 16KiB
Your original problem here was that that last debug-line (about how much was being uploaded) was itself not included in the web-paste. Did you figure out whether it is `fi.size()` that is wrong, here, or something else? Seeking to 48KiB and then reading 17KiB will get you at least the 16KiB requested -- but it's up to you what it does exactly, as long as it's documented :)

deprov447 replied:
actually I found we just have to refresh QFileInfo fi. Now uploads the log upto "_Uploading x bytes of logs_". 
![image](https://user-images.githubusercontent.com/57551996/113255068-05f46180-92e5-11eb-9840-b429603ac6b3.png)

adriaandegroot reviewed (commented):
Also needs 1 pass with the coding style tool

deprov447 committed:
[logUpload] suggestionsAndFixes-part3

- Resolved the problem of incomplete log upload
- sizeLimit = 0 fixed (turns off paste functionality)
- Documentation update
- sizeLimit < 0 now needs no hardcoded upper limit
- Calamares::Branding::uploadServerFromMap() serves sizeLimit in bytes

deprov447 committed:
[logUpload] Ran styleScript

deprov447 committed:
[logUpload] Added some basic tests

deprov447 committed:
[logUpload] Added one more test
